### PR TITLE
feat(dashboards): implement an endpoint to handle batch sharing of dashboards

### PR DESCRIPTION
**Pull request by jeremyjaouen**
## Description

This PR intends to add an endpoint to update all the shares of a dashboard.

The Dashboard Shares works as follow:
- An administrator can update shares of any users / contactgroups on any dashboards
- A non administrator  can update shares only if he has Dashboards Write ACLs and the dashboard is share with him as editor
- A non administrator can only update shares of contacts members of his access groups
- A non administrator can only update shares of contactgroups he is member of.

### Request body example: (PUT /configuration/dashboards/1/shares)

```json
{
  "contacts" : [{"id":  20, "role":  "viewer"},{"id":  21, "role":  "editor"}],
  "contact_groups" : [{"id": 6, "role":  "viewer"}]
}
```

### Tests
image

**Fixes** # MON-34236

## Type of change

- [ ] Patch fixing an issue (non-breaking change)
- [x] New functionality (non-breaking change)
- [ ] Breaking change (patch or feature) that might cause side effects breaking part of the Software

## Target serie

- [ ] 22.04.x
- [ ] 22.10.x
- [ ] 23.04.x
- [x] 23.10.x
- [x] 24.04.x (master)

#### Community contributors & Centreon team

- [x] I have followed the **coding style guidelines** provided by Centreon
- [x] I have commented my code, especially new **classes**, **functions** or any **legacy code** modified. (***docblock***)
- [x] I have commented my code, especially **hard-to-understand areas** of the PR.
- [x] I have **rebased** my development branch on the base branch (master, maintenance).

### CODE REVIEW

File: centreon/src/Core/Contact/Application/Repository/ReadContactRepositoryInterface.php [hunk 1]
```diff
@@ -89,4 +98,14 @@ public function existInAccessGroups(int $contactId, array $accessGroupIds): bool
      * @return Contact[]
      */
     public function findAdminWithRequestParameters(RequestParametersInterface $requestParameters): array;
+
+    /**
+     * @param int[] $contactIds
```
Comment (a-launois) [inline centreon/src/Core/Contact/Application/Repository/ReadContactRepositoryInterface.php:103]: a description of what does the method do would be nice

File: centreon/src/Core/Dashboard/Application/UseCase/ShareDashboard/ShareDashboard.php [hunk 2]
```diff
@@ -0,0 +1,202 @@
+<?php
+
+/*
+ * Copyright 2005 - 2023 Centreon (https://www.centreon.com/)
```
Comment (a-launois) [inline centreon/src/Core/Dashboard/Application/UseCase/ShareDashboard/ShareDashboard.php:4]: Applicable for all new files:
```suggestion
 * Copyright 2005 - 2024 Centreon (https://www.centreon.com/)
```

File: centreon/src/Core/Dashboard/Infrastructure/API/ShareDashboard/ShareDashboardRoute.yaml [hunk 3]
```diff
@@ -0,0 +1,7 @@
+ShareDashboardPlaylist:
+  methods: PUT
+  path: /configuration/dashboards/{dashboardId}/shares
+  requirements:
+    dashboardId: '\d+'
+  controller: 'Core\Dashboard\Infrastructure\API\ShareDashboard\ShareDashboardController'
+  condition: "request.attributes.get('version') >= 23.10"
```
Comment (a-launois) [inline centreon/src/Core/Dashboard/Infrastructure/API/ShareDashboard/ShareDashboardRoute.yaml:7]:
```suggestion
  condition: "request.attributes.get('version') >= 24.10"
```

**Review by a-launois — CHANGES_REQUESTED** [review]

File: centreon/src/Core/Dashboard/Application/UseCase/ShareDashboard/ShareDashboard.php (hunk 2 above)
Comment (jeremyjaouen) [reply centreon/src/Core/Dashboard/Application/UseCase/ShareDashboard/ShareDashboard.php:4]: Will try to update it but this change currently break the pipeline on modules, don't know if it is the case on web

File: centreon/src/Core/Dashboard/Infrastructure/API/ShareDashboard/ShareDashboardRoute.yaml (hunk 3 above)
Comment (jeremyjaouen) [reply centreon/src/Core/Dashboard/Infrastructure/API/ShareDashboard/ShareDashboardRoute.yaml:7]: 23.10 is correct as this new endpoint is also backported on 23.10

**Review by a-launois — APPROVED** [review]

File: centreon/src/Core/Contact/Application/Repository/ReadContactRepositoryInterface.php [hunk 4]
```diff
@@ -89,4 +98,16 @@ public function existInAccessGroups(int $contactId, array $accessGroupIds): bool
...
+
+    /**
+     * Retrieve admins based on given IDs.
+     *
+     * @param int[] $contactIds
+     *
+     * @throws \Throwable
+     *
+     * @return Contact[]
+     *                   /
```
Comment (adr-mo) [inline centreon/src/Core/Contact/Application/Repository/ReadContactRepositoryInterface.php:110]: petit loupé

Comment (jeremyjaouen) [reply centreon/src/Core/Contact/Application/Repository/ReadContactRepositoryInterface.php:110]: mdr

**Review by adr-mo — APPROVED** [review]

**Pull request merged**